src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -2927,8 +3056,11 @@ public override void VisitForEachIterationVariables(BoundForEachStatement node)
                     TypeSymbolWithAnnotations destinationType = iterationVariable.Type;
                     HashSet<DiagnosticInfo> useSiteDiagnostics = null;
                     Conversion conversion = _conversions.ClassifyImplicitConversionFromType(sourceType.TypeSymbol, destinationType.TypeSymbol, ref useSiteDiagnostics);
-                    // PROTOTYPE(NullableReferenceTypes): Report nullability mismatch if any.
-                    TypeSymbolWithAnnotations result = InferResultNullability(operand: null, conversion, destinationType.TypeSymbol, sourceType);
+                    TypeSymbolWithAnnotations result = ApplyConversion(node.IterationVariableType, operandOpt: null, conversion, destinationType.TypeSymbol, sourceType, checkConversion: false, fromExplicitCast: true, out bool canConvert);
+                    if (destinationType.IsReferenceType && destinationType.IsNullable == false && sourceType.IsNullable == true)
+                    {
+                        ReportStaticNullCheckingDiagnostics(ErrorCode.WRN_ConvertingNullableToNonNullable, node.IterationVariableType.Syntax);
+                    }
```

Comment:
There are multiple conversions involved in a foreach statement, based on looking at the lowering logic. Are they all properly checked for nullability?

```
        /// E e = ((C)(x)).GetEnumerator()
        /// try {
        ///     while (e.MoveNext()) {
        ///         V v = (V)(T)e.Current;
```

Reply:
We're not checking the implicit conversion to `(T)` although it's not clear that would produce a warning.

---
In reply to: [189381113](https://github.com/dotnet/roslyn/pull/26356#discussion_r189381113) [](ancestors = 189381113)

Reply:
Please add test or PROTOTYPE. Thanks

---
In reply to: [189414135](https://github.com/dotnet/roslyn/pull/26356#discussion_r189414135) [](ancestors = 189414135,189381113)